test/test_python_dispatch.py

```diff
@@ -335,6 +336,34 @@ def backward(ctx, grad_output):
 $5 = torch._ops.aten.mul($4, $0)
 $6 = torch._ops.aten.add_($1, $5)''')
 
+    def test_subclass_creation(self):
+        # Make sure these statements runs without error
+        # In particular checking that when internal detach returns
+        # subclasses, these are cleanly overwritten.
+        class Foo(torch.Tensor):
+            pass
+        a = torch.Tensor._make_subclass(Foo, LoggingTensor(torch.rand(2)))
+        b = LoggingTensor(torch.rand(2)).as_subclass(Foo)
```

Comment:
Thinking about the preferred behavior here, I kind of want `as_subclass` to hit up `__torch_function__` and the default implementation to raise an error if Foo isn't a subclass of LoggingTensor. Because you have no reason to believe that slicing off all of the LoggingTensor state and turning it into a Foo is a valid operation.

Reply:
If we go to `__torch_function__`, how is this default implementation supposed to work? Because even if it is a proper subclass, we will still have the problem that the `.detach()` will generate a Tensor with a pyobj already (of the parent class).
And so if we refuse to "replace" these pyobject, we can't really make such a subclass?

Also, we have some subclass (like nn.Parameter) that disable `__torch_function__` so that might be problematic.